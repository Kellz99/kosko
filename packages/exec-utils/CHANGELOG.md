# @kosko/exec-utils

## 0.1.0
### Minor Changes



- [#90](https://github.com/tommy351/kosko/pull/90) [`5b71844`](https://github.com/tommy351/kosko/commit/5b71844e700fdec9225c6c1395004a12e0869254) Thanks [@tommy351](https://github.com/tommy351)! - First release.
