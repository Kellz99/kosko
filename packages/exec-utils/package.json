{
  "name": "@kosko/exec-utils",
  "version": "0.2.0",
  "description": "Utilities for executing external processes.",
  "homepage": "https://kosko.dev",
  "repository": "https://github.com/tommy351/kosko/tree/master/packages/exec-utils",
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kubernetes",
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=14"
  },
  "dependencies": {
    "cross-spawn": "^7.0.3",
    "tslib": "^2.4.0"
  },
  "devDependencies": {
    "@kosko/jest-preset": "workspace:^0.0.0",
    "@types/cross-spawn": "^6.0.2",
    "get-stdin": "^8.0.0"
  },
  "jest": {
    "preset": "@kosko/jest-preset"
  }
}
