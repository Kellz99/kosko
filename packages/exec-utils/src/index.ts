export * from "./arg";
export * from "./error";
export * from "./spawn";
