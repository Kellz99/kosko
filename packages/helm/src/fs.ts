import { getErrorCode } from "@kosko/common-utils";
import { cp, rename, rm, stat } from "node:fs/promises";

export async function move(src: string, dest: string): Promise<void> {
  try {
    await rename(src, dest);
  } catch (err) {
    if (getErrorCode(err) !== "EXDEV") throw err;

    await cp(src, dest, {
      recursive: true,
      errorOnExist: true,
      preserveTimestamps: true
    });
    await rm(src, { recursive: true, force: true });
  }
}

export async function fileExists(path: string): Promise<boolean> {
  try {
    const stats = await stat(path);
    return stats.isFile();
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;
    return false;
  }
}
