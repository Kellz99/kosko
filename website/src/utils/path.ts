export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
