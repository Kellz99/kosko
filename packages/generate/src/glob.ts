import { readdir } from "fs/promises";
import { join, posix } from "path";
import mm from "micromatch";

function createMatcher(patterns: readonly string[], options?: mm.Options) {
  const matchers = patterns.map((pattern) => mm.matcher(pattern, options));
  return (path: string): boolean => matchers.some((matcher) => matcher(path));
}

export interface GlobOptions {
  path: string;
  patterns: readonly string[];
  extensions: readonly string[];
}

export interface GlobResult {
  relativePath: string;
  absolutePath: string;
}

export async function* glob(options: GlobOptions): AsyncIterable<GlobResult> {
  const matcherOptions: mm.Options = { cwd: options.path };
  const suffix =
    options.extensions.length > 1
      ? `.{${options.extensions.join(",")}}`
      : `.${options.extensions.join(",")}`;
  const fileMatcher = createMatcher(
    options.patterns.map((pattern) => pattern + suffix),
    matcherOptions
  );
  const dirMatcher = createMatcher(options.patterns, matcherOptions);

  async function* walk(
    path: string,
    parent: string
  ): AsyncIterable<GlobResult> {
    for (const file of await readdir(path, { withFileTypes: true })) {
      const result: GlobResult = {
        relativePath: posix.join(parent, file.name),
        absolutePath: join(path, file.name)
      };

      if (file.isDirectory()) {
        if (dirMatcher(result.relativePath)) {
          yield result;
        }

        yield* walk(result.absolutePath, result.relativePath);
      } else if (fileMatcher(result.relativePath)) {
        yield result;
      }
    }
  }

  yield* walk(options.path, "");
}
