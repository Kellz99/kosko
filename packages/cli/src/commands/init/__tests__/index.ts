/// <reference types="jest-extended"/>
import { InitArguments, initCmd } from "../index";
import tempDir from "temp-dir";
import tmp from "tmp-promise";
import fs from "fs";
import { join, posix } from "path";
import glob from "fast-glob";
import { spawn } from "@kosko/exec-utils";

let tmpDir: tmp.DirectoryResult;

jest.mock("@kosko/log");
jest.mock("@kosko/exec-utils");

async function execute(args: Partial<InitArguments>): Promise<void> {
  await initCmd.handler(args as any);
}

async function listAllFiles(dir: string): Promise<Record<string, string>> {
  const paths = await glob("**/*", { cwd: dir });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await fs.promises.readFile(
      join(dir, path),
      "utf8"
    );
  }

  return files;
}

beforeEach(() => jest.resetAllMocks());

beforeEach(async () => {
  tmpDir = await tmp.dir({ tmpdir: tempDir, unsafeCleanup: true });
});

afterEach(() => tmpDir.cleanup());

describe("when the target exists and is not a directory", () => {
  let tmpFile: tmp.FileResult;

  beforeEach(async () => {
    tmpFile = await tmp.file({ tmpdir: tempDir });
  });

  afterEach(() => tmpFile.cleanup());

  test("should throw an error", async () => {
    await expect(execute({ path: tmpFile.path })).rejects.toThrow(
      "Path is not a directory"
    );
  });
});

describe("when the target exists and is not empty", () => {
  beforeEach(async () => {
    await fs.promises.writeFile(join(tmpDir.path, "test"), "foobar");
  });

  test("should throw an error without --force option", async () => {
    await expect(execute({ path: tmpDir.path })).rejects.toThrow(
      "Path already exists"
    );
  });

  test("should proceed with --force option", async () => {
    await expect(execute({ path: tmpDir.path, force: true })).toResolve();
  });

  test("should update package.json", async () => {
    const packageJsonPath = join(tmpDir.path, "package.json");

    await fs.promises.writeFile(
      packageJsonPath,
      JSON.stringify(
        {
          name: "foo",
          version: "1.2.3",
          dependencies: {
            debug: "3.2.1"
          }
        },
        null,
        "  "
      )
    );
    await expect(execute({ path: tmpDir.path, force: true })).toResolve();
    await expect(
      fs.promises.readFile(packageJsonPath, "utf8")
    ).resolves.toMatchSnapshot();
  });
});

describe("when the target exists and existing files can be ignored", () => {
  beforeEach(async () => {
    await fs.promises.writeFile(join(tmpDir.path, ".DS_Store"), "");
  });

  test("should succeed", async () => {
    await expect(execute({ path: tmpDir.path })).toResolve();
  });

  test("should throw an error if other unignorable files exist", async () => {
    await fs.promises.writeFile(join(tmpDir.path, "test"), "foobar");
    await expect(execute({ path: tmpDir.path })).rejects.toThrow(
      "Path already exists"
    );
  });
});

describe("when the target exists and only git folder exist", () => {
  beforeEach(async () => {
    await fs.promises.mkdir(join(tmpDir.path, ".git"));
  });

  test("should succeed", async () => {
    await expect(execute({ path: tmpDir.path })).toResolve();
  });
});

describe("when the target exists and only contain log files", () => {
  beforeEach(async () => {
    await fs.promises.writeFile(join(tmpDir.path, "npm.log"), "foo");
  });

  test("should succeed", async () => {
    await expect(execute({ path: tmpDir.path })).toResolve();
  });
});

describe("when path is not specified", () => {
  test("should use cwd instead", async () => {
    await expect(execute({ cwd: tmpDir.path })).toResolve();
  });
});

describe("when relative path is given", () => {
  test("should init in a new folder", async () => {
    await execute({ cwd: tmpDir.path, path: "foo" });

    const stats = await fs.promises.stat(join(tmpDir.path, "foo"));
    expect(stats.isDirectory()).toBeTrue();
  });
});

describe("when no options are given", () => {
  beforeEach(async () => {
    await execute({ path: tmpDir.path });
  });

  test("should generate files", async () => {
    await expect(listAllFiles(tmpDir.path)).resolves.toMatchSnapshot();
  });

  test("should not run install", async () => {
    expect(spawn).not.toHaveBeenCalled();
  });
});

describe("when --install option is given", () => {
  test("should run install in the folder", async () => {
    await execute({ path: tmpDir.path, install: true });
    expect(spawn).toHaveBeenCalledTimes(1);
    expect(spawn).toHaveBeenCalledWith("npm", ["install"], {
      cwd: tmpDir.path,
      stdio: "inherit",
      env: {
        ...process.env,
        ADBLOCK: "1",
        DISABLE_OPENCOLLECTIVE: "1"
      }
    });
  });
});

describe("when --esm option is given", () => {
  test("should generate files", async () => {
    await execute({ path: tmpDir.path, esm: true });
    await expect(listAllFiles(tmpDir.path)).resolves.toMatchSnapshot();
  });
});

describe("when --typescript option is given", () => {
  test("should generate files", async () => {
    await execute({ path: tmpDir.path, typescript: true });
    await expect(listAllFiles(tmpDir.path)).resolves.toMatchSnapshot();
  });
});

describe("when --typescript and --esm option is given", () => {
  test("should generate files", async () => {
    await execute({ path: tmpDir.path, typescript: true, esm: true });
    await expect(listAllFiles(tmpDir.path)).resolves.toMatchSnapshot();
  });
});
