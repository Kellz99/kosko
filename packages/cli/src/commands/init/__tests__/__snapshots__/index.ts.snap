// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when --esm option is given should generate files 1`] = `
Object {
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
npm run generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
npm run --silent generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
npm run generate -- --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.js": "import { Deployment } from \\"kubernetes-models/apps/v1/Deployment\\";
import { Service } from \\"kubernetes-models/v1/Service\\";
import env from \\"@kosko/env\\";

const params = await env.component(\\"nginx\\");

const name = \\"nginx\\";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: \\"nginx\\",
            image: \\"nginx:stable\\",
            ports: [{ containerPort: port }],
            env: [{ name: \\"NGINX_PORT\\", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

export default [deployment, service];
",
  "environments/dev/index.js": "export default {
  namespace: \\"dev\\"
};
",
  "environments/dev/nginx.js": "export default {
  replicas: 3
};
",
  "kosko.toml": "components = [\\"*\\"]
",
  "package.json": "{
  \\"scripts\\": {
    \\"generate\\": \\"kosko generate\\",
    \\"validate\\": \\"kosko validate\\"
  },
  \\"dependencies\\": {
    \\"@kosko/env\\": \\"^3.0.0\\",
    \\"kosko\\": \\"^2.0.0\\",
    \\"kubernetes-models\\": \\"^3.0.0\\"
  },
  \\"type\\": \\"module\\"
}",
}
`;

exports[`when --typescript and --esm option is given should generate files 1`] = `
Object {
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
npm run generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
npm run --silent generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
npm run generate -- --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.ts": "import { Deployment } from \\"kubernetes-models/apps/v1/Deployment\\";
import { Service } from \\"kubernetes-models/v1/Service\\";
import env from \\"@kosko/env\\";

const params = await env.component(\\"nginx\\");

const name = \\"nginx\\";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: \\"nginx\\",
            image: \\"nginx:stable\\",
            ports: [{ containerPort: port }],
            env: [{ name: \\"NGINX_PORT\\", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

export default [deployment, service];
",
  "env.d.ts": "import \\"@kosko/env\\";

// Declare types for global environment variables
declare interface GlobalEnvironment {
  namespace: string;
}

// Declare types for component environment variables
declare interface ComponentEnvironment {
  // Fallback type of all other component variables which are not specified below
  [key: string]: unknown;

  // Specify types for each component
  nginx: {
    replicas: number;
  };
}

declare module \\"@kosko/env\\" {
  interface Environment {
    global(): GlobalEnvironment;

    component<K extends string>(
      name: K
    ): GlobalEnvironment & ComponentEnvironment[K];
  }
}
",
  "environments/dev/index.ts": "export default {
  namespace: \\"dev\\"
};
",
  "environments/dev/nginx.ts": "export default {
  replicas: 3
};
",
  "kosko.toml": "components = [\\"*\\"]
require = [\\"ts-node/register\\"]
",
  "package.json": "{
  \\"scripts\\": {
    \\"generate\\": \\"kosko generate\\",
    \\"validate\\": \\"kosko validate\\"
  },
  \\"dependencies\\": {
    \\"@kosko/env\\": \\"^3.0.0\\",
    \\"kosko\\": \\"^2.0.0\\",
    \\"kubernetes-models\\": \\"^3.0.0\\"
  },
  \\"devDependencies\\": {
    \\"@tsconfig/recommended\\": \\"^1.0.1\\",
    \\"ts-node\\": \\"^10.4.0\\",
    \\"typescript\\": \\"^4.5.3\\"
  },
  \\"type\\": \\"module\\"
}",
  "tsconfig.json": "{
  \\"extends\\": \\"@tsconfig/recommended/tsconfig.json\\",
  \\"compilerOptions\\": {
    \\"module\\": \\"esnext\\"
  }
}",
}
`;

exports[`when --typescript option is given should generate files 1`] = `
Object {
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
npm run generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
npm run --silent generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
npm run generate -- --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.ts": "import { Deployment } from \\"kubernetes-models/apps/v1/Deployment\\";
import { Service } from \\"kubernetes-models/v1/Service\\";
import env from \\"@kosko/env\\";

const params = env.component(\\"nginx\\");

const name = \\"nginx\\";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: \\"nginx\\",
            image: \\"nginx:stable\\",
            ports: [{ containerPort: port }],
            env: [{ name: \\"NGINX_PORT\\", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

export default [deployment, service];
",
  "env.d.ts": "import \\"@kosko/env\\";

// Declare types for global environment variables
declare interface GlobalEnvironment {
  namespace: string;
}

// Declare types for component environment variables
declare interface ComponentEnvironment {
  // Fallback type of all other component variables which are not specified below
  [key: string]: unknown;

  // Specify types for each component
  nginx: {
    replicas: number;
  };
}

declare module \\"@kosko/env\\" {
  interface Environment {
    global(): GlobalEnvironment;

    component<K extends string>(
      name: K
    ): GlobalEnvironment & ComponentEnvironment[K];
  }
}
",
  "environments/dev/index.ts": "export default {
  namespace: \\"dev\\"
};
",
  "environments/dev/nginx.ts": "export default {
  replicas: 3
};
",
  "kosko.toml": "components = [\\"*\\"]
require = [\\"ts-node/register\\"]
",
  "package.json": "{
  \\"scripts\\": {
    \\"generate\\": \\"kosko generate\\",
    \\"validate\\": \\"kosko validate\\"
  },
  \\"dependencies\\": {
    \\"@kosko/env\\": \\"^3.0.0\\",
    \\"kosko\\": \\"^2.0.0\\",
    \\"kubernetes-models\\": \\"^3.0.0\\"
  },
  \\"devDependencies\\": {
    \\"@tsconfig/recommended\\": \\"^1.0.1\\",
    \\"ts-node\\": \\"^10.4.0\\",
    \\"typescript\\": \\"^4.5.3\\"
  }
}",
  "tsconfig.json": "{
  \\"extends\\": \\"@tsconfig/recommended/tsconfig.json\\"
}",
}
`;

exports[`when no options are given should generate files 1`] = `
Object {
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
npm run generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
npm run --silent generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
npm run generate -- --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.js": "const { Deployment } = require(\\"kubernetes-models/apps/v1/Deployment\\");
const { Service } = require(\\"kubernetes-models/v1/Service\\");
const env = require(\\"@kosko/env\\");

const params = env.component(\\"nginx\\");

const name = \\"nginx\\";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: \\"nginx\\",
            image: \\"nginx:stable\\",
            ports: [{ containerPort: port }],
            env: [{ name: \\"NGINX_PORT\\", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

module.exports = [deployment, service];
",
  "environments/dev/index.js": "module.exports = {
  namespace: \\"dev\\"
};
",
  "environments/dev/nginx.js": "module.exports = {
  replicas: 3
};
",
  "kosko.toml": "components = [\\"*\\"]
",
  "package.json": "{
  \\"scripts\\": {
    \\"generate\\": \\"kosko generate\\",
    \\"validate\\": \\"kosko validate\\"
  },
  \\"dependencies\\": {
    \\"@kosko/env\\": \\"^3.0.0\\",
    \\"kosko\\": \\"^2.0.0\\",
    \\"kubernetes-models\\": \\"^3.0.0\\"
  }
}",
}
`;

exports[`when the target exists and is not empty should update package.json 1`] = `
"{
  \\"name\\": \\"foo\\",
  \\"version\\": \\"1.2.3\\",
  \\"dependencies\\": {
    \\"@kosko/env\\": \\"^3.0.0\\",
    \\"debug\\": \\"3.2.1\\",
    \\"kosko\\": \\"^2.0.0\\",
    \\"kubernetes-models\\": \\"^3.0.0\\"
  },
  \\"scripts\\": {
    \\"generate\\": \\"kosko generate\\",
    \\"validate\\": \\"kosko validate\\"
  }
}"
`;
